File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        
        int first=nums.size()-1;
        
        while(first >= 1 && nums[first-1] >= nums[first]){
            first--;
        }
        
        if(first==0){
            sort(nums.begin(), nums.end());
            return;
        }
        
        first--;
        
        int second = nums.size()-1;
        while(first<second && nums[first] >= nums[second]){
            second--;
        }
        
        swap(nums[first], nums[second]);
        reverse(nums.begin()+first+1, nums.end());
    }
};

File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        
        first = len(nums)-1
        
        while first >= 1 and nums[first-1] >= nums[first]:
            first -= 1
        
        if first == 0:
            nums.sort()
            return
        
        first -= 1
        
        second = len(nums)-1
        
        while first < second and nums[first] >= nums[second]:
            second -= 1
            
        nums[first], nums[second] = nums[second], nums[first]
        
        start = first+1
        end = len(nums)-1
        
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
           
            start += 1
            end -= 1
        
        